plasmapy/classes/species.py

```diff
@@ -0,0 +1,231 @@
+"""
+Class representing a group of particles"""
+# coding=utf-8
+import numpy as np
+from scipy.interpolate import RegularGridInterpolator
+from plasmapy.constants import atomic
+from astropy import constants
+from astropy import units as u
+
+
+class Species:
+    """
+    Object representing a species of particles: ions, electrons, or simply
+    a group of particles with a particular initial velocity distribution.
+
+    Parameters
+    ----------
+    plasma : `Plasma`
+        plasma from which fields can be pulled
+    type : str
+        particle type. See `plasmapy.constants.atomic` for suitable arguments.
+        The default is a proton.
+    n: int
+        number of macroparticles. The default is a single particle.
+    scaling : float
+        number of particles represented by each macroparticle.
+        The default is 1, which means a 1:1 correspondence between particles
+        and macroparticles.
+    dt : `astropy.units.Quantity`
+        length of timestep
+    nt : int
+        number of timesteps
+
+    Attributes
+    ----------
+    x: `astropy.units.Quantity`
+    v: `astropy.units.Quantity`
+        Current position and velocity, respectively. Shape (n, 3).
+    position_history: `astropy.units.Quantity`
+    velocity_history: `astropy.units.Quantity`
+        History of position and velocity. Shape (nt, n, 3).
+    q: `astropy.units.Quantity`
+    m: `astropy.units.Quantity`
+        Charge and mass of particle.
+    eff_q: `astropy.units.Quantity`
+    eff_m: `astropy.units.Quantity`
+        Total charge and mass of macroparticle.
+    kinetic_energy
+    kinetic_energy_history
+
+    Examples
+    ----------
+    See `plasmapy/examples/particle-stepper.ipynb.`
+
+    """
+    def __init__(self, plasma, particle_type='p', n=1, scaling=1,
+                 dt=np.inf, nt=np.inf):
+        self.q = atomic.charge_state(particle_type) * constants.e.si
+        self.m = atomic.ion_mass(particle_type)
+        self.N = int(n)
+        self.scaling = scaling
+        self.eff_q = self.q * scaling
+        self.eff_m = self.m * scaling
+
+        self.plasma = plasma
+        if np.isinf(dt) and np.isinf(nt):
+            raise ValueError("Both dt and nt are infinite.")
+
+        self.dt = dt
+        assert self.dt.si.unit == u.s
+        self.NT = int(nt)
+        self.t = np.arange(nt) * dt
+
+        self.x = np.zeros((n, 3), dtype=float) * u.m
+        self.v = np.zeros((n, 3), dtype=float) * (u.m / u.s)
```

Comment:
I'm not quite sure. The way I'm doing this is that `x` and `v` do all the live calculation, so they're small arrays that can be kept in memory, while `position or velocity_history` do all the memory storage. This way, it's easier to convert the history into, say, a HDF5 dataset. I'll port that code a bit later.